### PR TITLE
build: namespace generated headers with `zephyr/`

### LINKED ISSUE #60377: posix: uname: follow up enhancements
Follow up enhancements to the `uname` PR in #59981
- https://github.com/zephyrproject-rtos/zephyr/pull/59981#discussion_r1257471696
- https://github.com/zephyrproject-rtos/zephyr/pull/59981#issuecomment-1627699316#63973 

**EDIT:**
This issue is about the namespacing of the generated `version.h`

### LINKED ISSUE #68035: [RFC] build: namespace generated headers with `zephyr/`
## Introduction

Two years ago there was a header migration in #45410 to namespace all the headers in the `include` directory to `include/zephyr`, however the build-time generated headers are (still) not namespaced, i.e.:

- `app_version.h`
- `cmake_intdef.h`
- `core-isa-dM.h`
- `devicetree_generated.h`
- `driver-validation.h`
- `kobj-types-enum.h`
- `linker-kobject-prebuilt-data.h`
- `mcuboot_version.h`
- `offsets.h`
- `otype-to-size.h`
- `otype-to-str.h`
- `strerror_table.h`
- `strsignal_table.h`
- `syscall_list.h`
- `version.h`
- `zsr.h`
- Kconfig headers (`autoconf.h`)
- and all the syscalls headers

Generated headers, especially `version.h`, can easily conflict with other libraries & user applications. This RFC is proposing to namespace the generated headers into the `zephyr/` directory as well.

### Problem description

Some of the generated headers have pretty generic names, i.e.: `version.h`, and can potentially conflict with other library / user applications. Ideally, all Zephyr's headers, in-tree or generated, should be namespaced with `zephyr/`.

### Proposed change

The generated headers previously in `build/zephyr/include/generated/` will now be placed in `build/zephyr/include/generated/zephyr/`, Zephyr sources that are including these headers will be updated accordingly, i.e.:

https://github.com/zephyrproject-rtos/zephyr/blob/114548ef607aefd712dda7e5ea954981e778185b/kernel/banner.c#L10

will be changed to 

```c
#include <zephyr/version.h>
```

## Detailed RFC

The task involves modifying CMake files and scripts to output generated headers in the `include/generated/zephyr/` directory, updating dependencies on these headers accordingly, changing include paths for in-tree sources, creating a compatibility Kconfig for uninterrupted user application builds, and updating migration and release notes with details on this change and its mitigation steps.

### Proposed change (Detailed)

- Modify CMake files and scripts to output headers to the `include/generated/zephyr/` directory
- Update CMake files and scripts to depend on headers in the `include/generated/zephyr/` directory
- Update in-tree sources' include paths to `zephyr/<header.h>`
- Create a compatibility Kconfig so that user applications will still continue to build
- Update the migration/release notes about the change, and steps to mitigate

### Dependencies

External libraries and user applications can be affected, in that case they may use the migration script.

   - [sof](https://github.com/thesofproject/sof)
     - https://github.com/thesofproject/sof/pull/8459
     - https://github.com/zephyrproject-rtos/sof/pull/34
   - [percepio](https://github.com/percepio/percepio)
     - https://github.com/percepio/percepio/pull/5
     - https://github.com/percepio/percepio/pull/6
     - https://github.com/zephyrproject-rtos/percepio/pull/10

### Concerns and Unresolved Questions

Can't think of any

## Alternatives

Rely on all external libaries & user applications to namespace their header so that no conflicts can occur.

### LINKED ISSUE #73114: [RFC] build: namespace generated headers with `zephyr/`
## Introduction

Related to #41543.
Multiple work has been done to move the in-tree header files into the `zephyr/` folder, however, the generated header files are still not namespaced with `zephyr/`, i.e.:

- `app_version.h`
- `cmake_intdef.h`
- `core-isa-dM.h`
- `devicetree_generated.h`
- `driver-validation.h`
- `kobj-types-enum.h`
- `linker-kobject-prebuilt-data.h`
- `mcuboot_version.h`
- `offsets.h`
- `otype-to-size.h`
- `otype-to-str.h`
- `strerror_table.h`
- `strsignal_table.h`
- `syscall_list.h`
- `version.h`
- `zsr.h`
- Kconfig headers (`autoconf.h`)
- and all the syscalls headers

### Problem description

Some of the generated headers have pretty generic names, i.e.: `version.h`, and can potentially conflict with other library / user applications. Ideally, all Zephyr's headers, in-tree or generated, should be namespaced with `zephyr/`.

### Proposed change

To avoid potential conflicts with other library or user application headers due to generic names, introduce a zephyr/ namespace for all Zephyr-generated and in-tree headers. This can be achieved by modifying the CMake files and scripts responsible for generating these headers to include the `zephyr/` prefix in their file names and updating the include paths in source files accordingly. 

Additionally, ensure that the legacy include paths remains compatible during this transition and update the migration and release notes with details about this change and any necessary mitigation steps for users.

## Detailed RFC

The task involves modifying CMake files and scripts to output generated headers in the `include/generated/zephyr/` directory, updating dependencies on these headers accordingly, changing include paths for in-tree sources, creating a compatibility Kconfig for uninterrupted user application builds, and updating migration and release notes with details on this change and its mitigation steps.

### Proposed change (Detailed)

- Modify CMake files and scripts to output headers to the `include/generated/zephyr/` directory
- Update CMake files and scripts to depend on headers in the `include/generated/zephyr/` directory
- Update in-tree sources' include paths to `zephyr/<header.h>`
- Create a compatibility Kconfig so that user applications will still continue to build
- Update the migration/release notes about the change, and steps to mitigate

### Dependencies

   - [sof](https://github.com/thesofproject/sof)
     - https://github.com/thesofproject/sof/pull/8459
     - https://github.com/zephyrproject-rtos/sof/pull/34
   - [percepio](https://github.com/percepio/percepio)
     - https://github.com/percepio/percepio/pull/5
     - https://github.com/percepio/percepio/pull/6
     - https://github.com/zephyrproject-rtos/percepio/pull/10

### Concerns and Unresolved Questions

N/A

## Alternatives

N/A

### PR DESCRIPTION
**Pull request by ycsin**
Namespaced the generated headers with `zephyr/` to prevent potential conflict with other headers.

Introduced a Kconfig (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH`) to support legacy generated header include paths in the mean time.

[Documentation preview](https://builds.zephyrproject.io/zephyr/pr/63973/docs/releases/migration-guide-3.7.html#build-system)

Fixes #60377
Fixes #68035

- [Architecture WG](https://github.com/zephyrproject-rtos/zephyr/pull/63973#issuecomment-1799139397)
- [x] Missing changes to the N repos that depend on this are required in order to merge the PR
   - [**sof**](https://github.com/thesofproject/sof)
     - https://github.com/thesofproject/sof/pull/8459
     - https://github.com/zephyrproject-rtos/sof/pull/34
   - [**percepio**](https://github.com/percepio/percepio)
     - https://github.com/percepio/percepio/pull/5
     - https://github.com/percepio/percepio/pull/6
     - https://github.com/zephyrproject-rtos/percepio/pull/10
     - https://github.com/percepio/percepio/pull/9
     - https://github.com/percepio/percepio/pull/8
     - https://github.com/zephyrproject-rtos/percepio/pull/13
   - https://github.com/zephyrproject-rtos/hal_stm32/pull/211
   - https://github.com/zephyrproject-rtos/hal_nxp/pull/395
   - https://github.com/zephyrproject-rtos/lvgl/pull/53
   - [**mcuboot**](https://github.com/mcu-tools/mcuboot)
     - https://github.com/mcu-tools/mcuboot/pull/1965
     - https://github.com/zephyrproject-rtos/mcuboot/pull/117

Automation script:

```python
#!/usr/bin/env python

from pathlib import Path
import re

EXTENSIONS = ("c", "h", "cpp", "rst", "md", "S", "ld", "py", "svg")

HEADERS = [
    "app_version",
    # "autoconf",
    "cmake_intdef",
    "core-isa-dM",
    "devicetree_generated",
    "driver-validation",
    "kobj-types-enum",
    "linker-kobject-prebuilt-data",
    "linker-kobject-prebuilt-priv-stacks",
    "linker-kobject-prebuilt-rodata",
    "mcuboot_version",
    "offsets",
    "otype-to-size",
    "otype-to-str",
    "strerror_table",
    "strsignal_table",
    "syscall_list",
    "version",
    "zsr",
]

HEADERS_H = [hdr + ".h" for hdr in HEADERS]

for p in Path(".").glob("zephyr/**/*"):
    if not p.is_file() or p.suffix and p.suffix[1:] not in EXTENSIONS:
        continue

    # skip git & github folders
    if (str(p).startswith("zephyr/.git")):
        continue

    # skip release docs
    if (str(p).startswith("zephyr/doc/releases")):
        continue

    content = ""
    with open(p, 'r', newline='', errors="surrogateescape") as f:
        for line in f:
            # for lines in sources
            m = re.match(r"^(.*)#include <(.*)>(.*)$", line)
            # for lines in docs
            n = re.match(r"^(.*)include/generated/(.*)\.h(.*)$", line)
            if (m and (m.group(2).startswith("syscalls/") or m.group(2) in HEADERS_H)):
                content += (m.group(1) + "#include <zephyr/" +
                            m.group(2) + ">" + m.group(3) + "\n")
            elif (n and n.group(2) in HEADERS):
                content += (n.group(1) + "include/generated/zephyr/" +
                            n.group(2) + ".h" + n.group(3) + "\n")
            else:
                content += line

    with open(p, "w", newline='') as f:
        f.write(content)

```

## [Requirements for Treewide Changes](https://docs.zephyrproject.org/latest/contribute/guidelines.html#requirements-for-treewide-changes)

- [x] The zephyr repository must apply the ‘treewide’ GitHub label to any issues or pull requests that are treewide changes
- [x] The person proposing a treewide change must create an [RFC issue](https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=RFC&template=003_rfc-proposal.md&title=) describing the change, its rationale and impact, etc. before any pull requests related to the change can be merged
- [x] The project’s [Architecture Working Group (WG)](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group) must include the issue on the agenda and discuss whether the project will accept or reject the change before any pull requests related to the change can be merged (with escalation to the TSC if consensus is not reached at the WG)
- [ ] The Architecture WG must specify the procedure for merging any PRs associated with each individual treewide change, including any required approvals for pull requests affecting specific subsystems or extra review time requirements
- [x] The person proposing a treewide change must email [devel@lists.zephyrproject.org](mailto:devel%40lists.zephyrproject.org) about the RFC if it is accepted by the Architecture WG before any pull requests related to the change can be merged

### CODE REVIEW

**Review by nordicjm — DISMISSED** [review]

File: CMakeLists.txt
Comment (gmarull) [inline CMakeLists.txt:557]: we should probably move all generated headers to `build/include/zephyr`, and include `build/include` (also `build/include/zephyr` to not break immediately).

Comment (gmarull) [reply CMakeLists.txt:557]: it would be guarded by a Kconfig option, enabled by default now, deprecated later maybe (same as we did for the public includes a while back)

**Comment by ycsin** [comment]
this PR is getting more involved than I initially expected

**Review by cfriedt — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@ycsin hey thanks for putting this together, since it's a major change I added the treewide and architecture review tags so this can be discussed in the architecture WG (cc @carlescufi)

**Review by fabiobaltieri — DISMISSED** [review]

**Review by de-nordic — DISMISSED** [review]
Flash and MCUmgr look ok.

File: Kconfig.zephyr
Comment (gmarull) [inline Kconfig.zephyr:918]: I'd avoid `select DEPRECATED` in a first iteration, some builds enable `WARN_DEPRECATED`.

**Review by finikorg — DISMISSED** [review]

**Comment by ycsin** [comment]
Rebased on top of `main`
Moved `LEGACY_GENERATED_INCLUDE_PATH` into the Compatibility menu

**Review by de-nordic — DISMISSED** [review]
OK for Flash API and MCUmgr.

File: CMakeLists.txt
Comment (gmarull) [inline CMakeLists.txt:113]: I'm fine with the PR, just one note for discussion (Arch WG): should generated headers have their own subfolder, ie, `<zephyr/generated/xxx.h>`? (maybe with some exceptions like `version.h`).

**Comment by ycsin** [comment]
Another compilation issue getting fixed @ #64251

File: CMakeLists.txt
Comment (ycsin) [reply CMakeLists.txt:113]: +1, had the same thing in my mind

**Review by tejlmand — DISMISSED** [review]
Thanks for this PR, i'm very much in favor of namespacing Zephyr headers.

That said, I see no justification for the changes to Kconfiglib.py, as kconfiglib.py is a project which originates outside of Zephyr.
Thus it should only be patched in Zephyr for good reasons, such as bug fixing or important feature extensions which cannot be done in other ways.

File: scripts/kconfig/kconfiglib.py
Comment (ycsin) [reply scripts/kconfig/kconfiglib.py:2931]: Or should we leave this file as-is instead and merge this with superpower? cc @nashif

File: CMakeLists.txt
Comment (gmarull) [reply CMakeLists.txt:113]: Some will always be generated (e.g. syscall stuff, devicetree, etc.). I think it makes sense to put them into `generated/`, others like `version.h`, probably not.

**Review by mbolivar-ampere — DISMISSED** [review]
Devicetree changes look great. Thank you for taking the time to update the documentation.

This +1 is limited to the DT changes.

File: scripts/kconfig/kconfiglib.py
Comment (tejlmand) [reply scripts/kconfig/kconfiglib.py:2931]: after you reverted the changes pointed out in this comment https://github.com/zephyrproject-rtos/zephyr/pull/63973#discussion_r1368619501

then it should be as simple as dropping this commit https://github.com/zephyrproject-rtos/zephyr/pull/63973/commits/6ca708fb1d735482a73212de6c1f819a0145ba66 because you are no longer modifying kconfiglib.py.
So therefore CI should still be green (not considering other potential failures) 😉

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
Ran this again the push CI job in the testing repo (https://github.com/zephyrproject-rtos/zephyr-testing/runs/18077860182) single failure but it's unrelated. Great job there, hopefully we can merge it soon, too bad we did not make to this at the last arch wg, we should probably do it early next time, I don't see any objections here anyway.

Needs a rebase btw.

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by ycsin** [comment]
rebased and remove modifications to `scripts/kconfig/kconfiglib.py`

**Review by de-nordic — DISMISSED** [review]
Flash API and MCUmgr look ok

**Comment by ycsin** [comment]
rebased again for #64464

**Review by tejlmand — DISMISSED** [review]
Approved, but please let the reviewer decide if review comments has been addressed (unless the reviewer has been unresponsive for an unreasonable long period).
image

It was less than 24h since the last comment from me was addressed here:
image

**Review by cfriedt — DISMISSED** [review]

**Comment by carlescufi** [comment]
Architecture WG:

- Checked that we do not have an applicable Kconfig option anymore that can be used, instead we will need the new `CONFIG_LEGACY_GENERATED_INCLUDE_PATH`. 
- Missing changes to the 2 repos that depend on this are required in order to merge the PR
- Missing a entry to the migration document

**Comment by cfriedt** [comment]
> this PR is getting more involved than I initially expected
> 
> EDIT: And CI is complaining about some lines that are not modified in this PR:

@ycsin - those compliance issues could also be fixed in a separate PR. This one will require a respin anyway, so that might be a good option

File: CMakeLists.txt
Comment (cfriedt) [reply CMakeLists.txt:113]: I guess the main concern is that a generated header could conflict with a static header, and that a generated header might incorrectly get included first.

Technically, that's still a risk today, so could be solved separately.

**Comment by carlescufi** [comment]
> I'm not sure which category should this change falls into in the migration guide, any idea?

build system

**Comment by henrikbrixandersen** [comment]
This needs a rebase. Is the target milestone still v3.6.0?

**Comment by ycsin** [comment]
> This needs a rebase. Is the target milestone still v3.6.0?

I'm not sure tbh, does this need to go thru an architecture review?

**Comment by henrikbrixandersen** [comment]
> I'm not sure tbh, does this need to go thru an architecture review?

Yes, it needs to follow the process described here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes

**Comment by teburd** [comment]
Please do continue the work here, +1 from on CI and arch wg agreement

**Comment by ycsin** [comment]
This PR should be ready for the architecture review now

**Comment by ycsin** [comment]
testing the changes for `autoconf.h`, list of files that includes the `autoconf.h`:

```
(.venv) ycsin@LAPTOP-ROG:~/zephyrproject$ find . -type f -exec grep -H '#include [<|"]autoconf\.h[>|"]' {} +
./modules/hal/stm32/scripts/tests/genllheaders/data/stm32_ll_tim.h:#include <autoconf.h>
./modules/hal/stm32/scripts/tests/genllheaders/data/stm32_ll_usart.h:#include <autoconf.h>
./modules/hal/stm32/scripts/genllheaders/header-template.j2:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fmpi2c.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_lptim.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_wwdg.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_icache.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_gpio.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_pwr.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_adc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_bdma.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_spi.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_utils.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_delayblock.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dlyb.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_lpuart.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_opamp.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_ipcc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_hsem.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_rcc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dmamux.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dac.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_i3c.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dma2d.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_i2c.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_system.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_comp.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_hrtim.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_swpmi.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_tim.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_crc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_cortex.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_ucpd.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_cordic.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dcache.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_sdmmc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_lpgpio.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fmac.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_usart.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_exti.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_rng.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_mdma.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_iwdg.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dma.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_bus.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_crs.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fmc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_pka.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_rtc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fsmc.h:#include <autoconf.h>
./modules/hal/nxp/s32/mcux/devices/S32K344/S32K344_features.h:#include "autoconf.h"
```

**Comment by ycsin** [comment]
Seems like it works this time, not sure what I missed last time for the `autoconf.h` changes. cc @carlescufi

**Review by cfriedt — DISMISSED** [review]

**Comment by ycsin** [comment]
rebased to fix conflicts

**Comment by ycsin** [comment]
The PR branchs in `percepio` are now merged while this branch isn't, which is likely to cause havoc to percepio users?

**Comment by nordicjm** [comment]
Needs rebase

**Review by cfriedt — DISMISSED** [review]
This needs a rebase. It would be good to get in prior to v3.7.0

**Comment by ycsin** [comment]
Removed `Architecture Review` as it had been reviewed a few months ago, it just requires approvals from the maintainers now.

File: doc/releases/migration-guide-3.7.rst
Comment (henrikbrixandersen) [inline doc/releases/migration-guide-3.7.rst:47]: Should these have a `.h` postfix?

File: Kconfig.zephyr
Comment (henrikbrixandersen) [inline Kconfig.zephyr:1056]: Shouldn't this be disabled by default? Otherwise we risk introducing new include statements using the old path.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: doc/releases/migration-guide-3.7.rst
Comment (ycsin) [reply doc/releases/migration-guide-3.7.rst:47]: The omission is actually intentional (implied by them being header files) but I can add them if you think it increases clarity

Comment (ycsin) [reply doc/releases/migration-guide-3.7.rst]: Fixed

File: Kconfig.zephyr
Comment (ycsin) [reply Kconfig.zephyr]: Removed `default y` and added `select DEPRECATED`

**Comment by ycsin** [comment]
Rebased and ran the script to update some new #include additions

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
@tejlmand Please revisit.

File: CMakeLists.txt
Comment (ycsin) [reply CMakeLists.txt:550]: @henrikbrixandersen I think the `default y` came from here

Comment (henrikbrixandersen) [reply CMakeLists.txt:550]: I see

**Comment by ycsin** [comment]
Things are failing now because `LEGACY_GENERATED_INCLUDE_PATH` is now disabled by default (previously it was enabled by default).

Currently I'm just trying to see if `hal_stm32` is the only one affected by the latest change, I think it probably makes sense to enable the `LEGACY_GENERATED_INCLUDE_PATH` by default so that this PR has less dependencies on the external modules, and disable + deprecate that after we have all PRs in external modules repo merged.

**Comment by ycsin** [comment]
- [ ] wait for CI to pass (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`)
- [ ] Remove last commit (4b03a50bcd49ca4bffc2accfda7283b04f317371) (
- [ ] Wait for CI to pass again (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=y`)

**Comment by henrikbrixandersen** [comment]
> - [ ] wait for CI to pass (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`)
> 
> - [ ] Remove last commit (4b03a50bcd49ca4bffc2accfda7283b04f317371)
> 
> - [ ] Wait for CI to pass again (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=y`)

If CI passes with `CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`, why revert it?

**Comment by ycsin** [comment]
> > * [ ]  wait for CI to pass (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`)
> > * [ ]  Remove last commit ([4b03a50](https://github.com/zephyrproject-rtos/zephyr/commit/4b03a50bcd49ca4bffc2accfda7283b04f317371))
> > * [ ]  Wait for CI to pass again (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=y`)
> 
> If CI passes with `CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`, why revert it?

The CI pass because the last commit updates the modules to point at my PRs. Correct me if I'm wrong, the `west.yml` will only be updated when these PRs are merged, so in the mean time we need to enable `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` until they are merged and `west.yml` updated.

I also agree with @gmarull that the downstream shouldn't break immediately when they sync to the latest `main` branch after this is merge, i.e. a nightly CI build that tracks the latest `main`. We will show them a nice warning message that asks them to update (while we wait for the `west.yml` to be updated anyway):

Screenshot 2024-05-24 at 2 55 39 PM

Once all upstream stuff have been updated to work properly, we can stop enabling the `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` by default and the downstream app will fail to compile, if they are still using the legacy path. They will have to enable the Kconfig explicitly and be presented with an additional deprecation warning.

**Comment by ycsin** [comment]
CI finally passed 🥳
Will revert the last commit now

File: CMakeLists.txt
Comment (marc-hb) [reply CMakeLists.txt:118]: Wait... there are _other_ people who look at build warnings? @henrikbrixandersen, you and I should start a club!

I've seen Kconfig and Device tree warnings pretty much every time I build anything more complex than samples/hello_world/... neither of these tools seems to have any `-Werror`, which in my experience is the only way to make anyone care.

Comment (marc-hb) [reply CMakeLists.txt:118]: The warning will be printed only for people who should fix their code ASAP so printing it every time looks good to me.

Comment (ycsin) [reply CMakeLists.txt:118]: Otherwise the `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` which is enabled by default now will just make everything passes silently. Most people might not notice that the headers are now relocated to `zephyr/` unless they looked at their build directory or the migration guide.

On the other extreme, if we disable `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` by default, the downstream apps will fail to build immediately, which I don't think is the right way to do in an OS of this scale.

The current implementation IMHO is the middle approach - we don't want to break downstream applications, but at the same time we want to make sure that developers are made aware of the changes, and given enough time to do their migration (as we do ours - some of the PRs aren't merged, `west.yml` not updated to point at them).

**Comment by ycsin** [comment]
@cfriedt @nordicjm @finikorg @fabiobaltieri @de-nordic @dcpleung @mbolivar-ampere 
This PR is now ready (again), could you please take another look again?

@gmarull this PR should align with #41543 and consistent with the header migration PR that you have made in the past (~v3.1.0?), could you please review?

**Review by jhedberg — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]
Changes to nRF QSPI NOR look ok.

**Review by cfriedt — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
This may use some coordination help from someone from @zephyrproject-rtos/release, plenty of dependencies, plus probably something will pop up after merge as well, worth keeping it in mind.

**Comment by ycsin** [comment]
ping assignee @tejlmand, PTAL, thanks!

**Comment by ycsin** [comment]
> ping assignee @tejlmand, PTAL, thanks!

another ping @tejlmand

**Review by nashif — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Comment by ycsin** [comment]
🥳🍀🤞

**Comment by marc-hb** [comment]
Just FYI https://github.com/thesofproject/sof/actions/runs/9278328176/job/25529164408

```
FileNotFoundError:  No such file or directory:
  '/zep_workspace/build-lnl/zephyr/include/generated/autoconf.h'
```

That's because one Python script in SOF extracts values from `autoconf.h` after the build.

The C compilation was fine and the CONFIG_LEGACY_GENERATED_INCLUDE_PATH warning was printed by CMake as expected.

Just FYI.

**Pull request merged**
